Netloc integration with mpich

KavithaTipturMadhu opened the pull request:
Features:
1. Integrate netloc topology query
2. network topology type classification
3. comm_split_type with hints: switch levels in tree topology

pavanbalaji commented on src/mpid/ch3/channels/nemesis/src/ch3_init.c:
```diff
@@ -44,6 +44,9 @@ static int split_type(MPIR_Comm * user_comm_ptr, int stype, int key,
         goto fn_exit;
+    } else if (stype == MPIX_COMM_TYPE_NEIGHBORHOOD) {
+        mpi_errno = MPIR_Comm_split_type_neighborhood(comm_ptr, stype, key, info_ptr, newcomm_ptr);
+        goto fn_exit;
```
You'll never hit this `else if` branch.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -0,0 +1,52 @@
+    MPIR_NETLOC_NETWORK_TYPE__FAT_TREE,
+    MPIR_NETLOC_NETWORK_TYPE__CLOS_NETWORK,
+    MPIR_NETLOC_NETWORK_TYPE__INVALID,
+} MPIR_Netloc_network_type;
```
This should be renamed to `MPIR_Netloc_network_topo_type`.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -0,0 +1,52 @@
+int MPIR_Netloc_parse_topology(netloc_topology_t topology,
+                               MPIR_Netloc_network_attributes * network_attr);
+
+int MPIR_get_network_end_point(MPIR_Netloc_network_attributes, netloc_topology_t netloc_topology,
```
First character after `MPIR_` should be capitalized.

KavithaTipturMadhu commented:
@pavanbalaji I have updated the patches as per review comments.

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
@raffenet @pavanbalaji Barring one freebsd test failing because of slurm, the tests are green and I have addressed all the comments. Could you guys give it a look?

pavanbalaji commented:
I've signed off on the first commit.  For the second commit, I made some cleanup changes and signed off on it.  Please take a look.

I'm not particularly happy with the third commit.  What's the point of the `node_index` structure?  It doesn't seem to be used anywhere.  Also, this patch is mixing the network type (e.g., InfiniBand) with the network topology (e.g., fat tree).  One can create a fat tree topology with other network types too.  I understand that you need some IB-specific code to figure out the edge weights (e.g., SDR, DDR, QDR), but I think this needs to be hidden inside netloc.  For example, can you have netloc add the "edge weight" for each network edge with the link bandwidth?  Then MPICH will not need any IB-specific code at the MPI layer and can simply rely on the edge weights to build the topology.

KavithaTipturMadhu commented:
Replying to your previous comment @pavanbalaji, the node_index struct was for torus network, I forgot to remove it. I'll remove that and double check for stray code. I have addressed the other comment to add bandwidth in netloc and used it. I have committed my netloc changes [here](https://github.com/pmodels/netloc/commit/47c6847c17a3fab86c7305b82309e3d69098c3e2)

raffenet commented:
@KavithaTipturMadhu should I review this or are the other comments still being addressed?

KavithaTipturMadhu commented:
@raffenet I have addressed the comments.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -0,0 +1,52 @@
+typedef struct {
+    netloc_node_t *node;
+    long index;
+} node_index;
```
I still see `node_index`.  Please really fix my comments.

KavithaTipturMadhu replied:
I am so sorry about that, I have fixed this.

KavithaTipturMadhu commented:
@raffenet @pavanbalaji I had missed one of the changes, I have pushed it now, sorry about that.

pavanbalaji commented on src/include/netloc_util.h:
```diff
@@ -0,0 +1,47 @@
+
+int MPIR_Get_network_end_point(MPIR_Netloc_network_attributes, netloc_topology_t netloc_topology,
+                               hwloc_topology_t hwloc_topology, hwloc_cpuset_t bindset,
+                               netloc_node_t ** end_point);
```
The function names are not consistent.  If it's related to netloc, can you use the `MPIR_Netloc_` prefix?

KavithaTipturMadhu replied:
This function is a utility I wrote to get netloc node for the current process. Does this name sound right? MPIR_Netloc_get_network_end_point

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -497,6 +619,65 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
+                                  &info_args_are_equal);
+
+    if (hintval_global != NULL)
+        MPL_free(hintval_global);
```
The checks to make sure the info hints are the same for all processes are being duplicated.  They need to be consolidated into one place.

pavanbalaji commented:
@raffenet I've not reviewed the algorithms in the above commits.  Can you do that?

raffenet commented:
Sure, I can review the algorithms.

KavithaTipturMadhu commented on src/mpi/init/netloc_util.c:
```diff
@@ -0,0 +1,465 @@
+                break;
+            }
+            netloc_get_all_edges(topology, node, &num_edges, &edges);
+            if (num_edges > 0) {
```
Removed the check

KavithaTipturMadhu commented on src/mpi/init/netloc_util.c:
```diff
@@ -0,0 +1,465 @@
+                        }
+                    }
+                    if (!added_previously) {
+                        traversal_order[end_index++] = neighbor;
```
Once add nodes are added, there is no more insertion. I have added assert to ensure end index is less than the number of nodes. Do you think that would suffice?

raffenet replied:
Yes, assert is a good idea. Thanks.

KavithaTipturMadhu commented on src/mpi/init/netloc_util.c:
```diff
@@ -0,0 +1,465 @@
+                netloc_get_all_edges(topology, topology->nodes[i], &num_edges, &edges);
+                for (j = 0; j < num_edges; j++) {
+                    if (network_attr->u.tree.node_levels[edges[j]->dest_node->__uid__] < 0) {
+                        traversal_order[traversal_end++] = edges[j]->dest_node;
```
I have moved the initialization statements to before the loop begins.

KavithaTipturMadhu commented:
I can attach a couple of raw netloc json files here to test manually but I don't have any automated positive tests for this. I have addressed other refactoring comments from Pavan.

raffenet commented:
OK, cool. I think I have a decent understanding of how the detection algorithms are supposed to work. I'm OK with moving forward with this.

KavithaTipturMadhu commented:
@raffenet I have attached here a couple of test json files here, mostly boundary conditions and some positive tests too. I made a couple of minor changes in code and would restart the tests. Could you take a look at the changes for me? Thanks.
[graph_examples.zip](https://github.com/pmodels/mpich/files/2107597/graph_examples.zip)

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

raffenet commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -393,6 +393,38 @@ static int compare_info_args(const void *sendbuf, void *recvbuf, int count, MPIR
+
+    /* Compare hintval string length */
+    mpi_errno =
+        reduce_info_hint(&hintval_size, &hintval_size_global, sizeof(int) / sizeof(char), comm_ptr,
```
The implementation of `reduce_info_hint` still seems a bit overly complex, but I think we can solve that later.

pavanbalaji replied:
Do we need the `reduce_info_hint` function?  Can't that just be embedded into this function since this is the only place that it's used?

raffenet commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -580,14 +759,8 @@ int MPIR_Comm_split_type_neighborhood(MPIR_Comm * comm_ptr, int split_type, int
-         * being returned to the user. */
-        *newcomm_ptr = NULL;
-    }
+    } else
```
A style nitpick - I prefer to leave the braces here. If we use braces for one part of a conditional, we should use them for all parts.

KavithaTipturMadhu replied:
Fixed this

raffenet commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -507,6 +629,63 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
+        goto fn_exit;
+    }
+#endif
+  use_node_comm:
```
Is this label accurate? It seems we are just returning a `NULL` comm and not doing anything with the node comm.

KavithaTipturMadhu replied:
Changed the name of the label

raffenet commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -12,6 +12,11 @@
 
+#ifdef HAVE_NETLOC
+#include "netloc_util.h"
+#include <math.h>
```
What's `math.h` for in this case?

KavithaTipturMadhu replied:
I had added it for torus split, removed it for this patch.

raffenet commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -493,6 +627,63 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
+        goto fn_exit;
+    }
+#endif
+  use_null_comm:
```
It might look better to set to `NULL` at the beginning of the function and then overwrite it later when conditions are met. Personally, all these `goto`s make me a little uncomfortable.

raffenet commented:
OK, I think my comments have been addressed. I have a couple minor fixups stashed locally (mostly to commit messages). @pavanbalaji should I signoff and merge?

pavanbalaji commented:
I'd recommend removing `min_subcomm_size` for now, till that's ready.  The current implementation for that is incorrect.  That can go into the next PR.

KavithaTipturMadhu commented:
@raffenet @pavanbalaji  Removed min_subcomm_size hint support, creating a new PR to support this.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -505,6 +589,57 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
+        /* if our bindset is not valid, skip topology-aware comm split */
+        if (!MPIR_Process.bindset_is_valid)
+            goto fn_exit;
+#endif
```
Is this needed for splitting by network topology?

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -579,14 +714,8 @@ int MPIR_Comm_split_type_neighborhood(MPIR_Comm * comm_ptr, int split_type, int
-         * like UNDEFINED.  This will result in MPI_COMM_NULL
-         * being returned to the user. */
-        *newcomm_ptr = NULL;
+        MPIR_Comm_split_type_network_topo(comm_ptr, key, info_ptr, newcomm_ptr);
```
This doesn't sound right.  If the user didn't give a valid info key, we call `MPIR_Comm_split_type_network_topo`?

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -579,14 +714,8 @@ int MPIR_Comm_split_type_neighborhood(MPIR_Comm * comm_ptr, int split_type, int
-        *newcomm_ptr = NULL;
+        MPIR_Comm_split_type_network_topo(comm_ptr, key, info_ptr, newcomm_ptr);
     }
-
```
White space change.

pavanbalaji commented:
Some of the commit messages need changes based on the removal of `subcomm_min_size`.

KavithaTipturMadhu commented:
@raffenet @pavanbalaji I have addressed all the comments, starting tests again.

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented:
You ignored this comment: `Some of the commit messages need changes based on the removal of subcomm_min_size.`

KavithaTipturMadhu commented:
@pavanbalaji sorry I missed that change, I have updated it. I also removed your sign-off as suggested.

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -542,42 +542,101 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
+
+    /*Atleast one of the processes has "nbhd_common_dirname" hint */
+    if (hintval_size_max) {
+        MPIR_Comm_split_type_nbhd_common_dir(comm_ptr, key, hintval, newcomm_ptr);
```
I'm not sure I understand this.  Why is this being called when `hintval_size_max != 0`?

KavithaTipturMadhu commented:
@pavanbalaji I have updated the pr as per your comments.

raffenet commented:
is this ready to merge?

KavithaTipturMadhu commented:
@raffenet i updated the patches with Pavan's comments. @pavanbalaji  do you want to take a look?

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -366,27 +366,68 @@ static int node_split_gpu_device(MPIR_Comm * comm_ptr, int key,
-static int compare_info_args(const void *sendbuf, void *recvbuf, int count, MPIR_Comm * comm_ptr,
-                             int *info_args_global)
+static int compare_info_hint(const char *hintval, MPIR_Comm * comm_ptr,
+                             int *arg_info_args_are_equal)
```
What's with the variable name?

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -542,42 +542,97 @@ int MPIR_Comm_split_type(MPIR_Comm * user_comm_ptr, int split_type, int key,
+    /* We plan on dispatching different NEIGHBORHOOD support to
+     * different parts of MPICH, based on the key provided in the
+     * info object.  Right now, the one NEIGHBORHOOD we support is
+     * "nbhd_common_dirname", implementation of which lives in ROMIO */
```
This comment is no longer relevant.

KavithaTipturMadhu replied:
Removed.

pavanbalaji commented on configure.ac:
```diff
@@ -1514,6 +1514,43 @@ HWLOC_DO_AM_CONDITIONALS
+
+if test "$have_hwloc" = "yes" ; then
+    if test "${with_netloc_prefix}" != "no" ; then
+        if test "${with_netloc_prefix}" != "sys"; then
```
Are we using `sys` for system-installed netloc instead of `system`?  Is that consistent with other uses?

KavithaTipturMadhu replied:
Sorry I missed this, updated to `system`.

pavanbalaji commented on configure.ac:
```diff
@@ -1514,6 +1514,43 @@ HWLOC_DO_AM_CONDITIONALS
+                                   [])],
+                          [AC_MSG_ERROR([the Netloc installation seems to be added from system path])], [])
+        fi
+        AC_DEFINE(HAVE_NETLOC,1,[Define if netloc is available])
```
`HAVE_NETLOC` is not defined for system-installed netloc?

KavithaTipturMadhu replied:
AC_DEFINE is after the if else block that checks for netloc, but within the `-with-netloc-prefix!=no` check. Hence, it will be defined for either system defined netloc or user specified path.

pavanbalaji commented:
Please have a prefix for each commit.

pavanbalaji commented:
You can also add my signoffs on the last two commits.

pavanbalaji commented:
The last commit should have "test" as the prefix

KavithaTipturMadhu commented:
@pavanbalaji I have addressed your comments.

pavanbalaji commented on src/mpi/comm/comm_split_type.c:
```diff
@@ -366,27 +366,67 @@ static int node_split_gpu_device(MPIR_Comm * comm_ptr, int key,
     MPIR_Errflag_t errflag = MPIR_ERR_NONE;
-    int info_args_are_equal = 0;
+    int info_args_global = 0;
+    int info_args_equal = 0;
```
Please use more descriptive variable names.

KavithaTipturMadhu replied:
I have renamed the variables and added a comment describing them

pavanbalaji commented:
You can add my signoff on `comm_split_type: refactoring comm_split_type_neighborhood`

pavanbalaji commented:
This commit needs a more detailed description: `test: update for nhbd matching info hint`

pavanbalaji commented:
This commit needs a more detailed description: `netloc: topology discovery and attributes`

pavanbalaji commented:
This commit has a totally random description.  Please be more careful when writing commit messages:

```
commit d70b41174d76bab0e51110871a37246b0720b959
Author: Kavitha Tiptur Madhu <kmadhu@anl.gov>
Date:   Tue May 22 22:24:34 2018 -0500

    comm_split_type: split by switch level for trees
    
    Signed-off-by: Pavan Balaji <balaji@anl.gov>
    
    comm split fixup
```

KavithaTipturMadhu commented:
Verified correctness of changes pushed by @pavanbalaji and merged into `comm_split_type: refactoring info argument comparison` patch.

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
test:jenkins/ch3/most
test:jenkins/ch4/most

KavithaTipturMadhu commented:
Two tests are failing because of slurm issues in freebsd. I checked the failing ucx test manually, seems to be working. Restarting ucx tests.

KavithaTipturMadhu commented:
test:jenkins/ch4/ucx

KavithaTipturMadhu commented:
Two tests in ucx were failing because of PMIx patches, informed @raffenet.
